### PR TITLE
Generate method takes eos_token_id

### PR DESCRIPTION
**Pull request by chandanms**
## Description
Added eos_token_id argument for generate method of Llama implementation. Now the generation stops when it encounters the eos_token_id

## Related Issue
Solves the issue #23 

## Motivation and Context
Without this, the generate only stops after hitting the max_token limit which produces incoherent sentences

## How Has This Been Tested?
Test script has been added to test_llama_implementation.py to test the generate method using dummy data

### CODE REVIEW

File: simple_stories_train/models/llama.py [hunk 1]
```diff
@@ -423,9 +424,13 @@ def generate(
             probs = F.softmax(logits, dim=-1)
             # sample from the distribution
             idx_next = torch.multinomial(probs, num_samples=1)
-            # append sampled index to the running sequence and continue
+            # append sampled index to the running sequence
             idx = torch.cat((idx, idx_next), dim=1)
 
+            # Check if EOS token was generated
+            if eos_token_id is not None and (idx_next == eos_token_id).any():
```
Comment (danbraunai) [inline simple_stories_train/models/llama.py:431]: How come the `.any()`? If this function works with batches, then this will stop generation if `idx_next == eos_token_id` for any sample, which is bad. In this case you'll probably want to just keep letting it generate until it hits the max tokens and later ignore everything after eos_token_id.

If the function doesn't work with batches, then you shouldn't need `.any()` (you'd also want to change the type hints in the function signature).

File: simple_stories_train/models/llama.py [hunk 2]
```diff
@@ -423,9 +424,13 @@ def generate(
             probs = F.softmax(logits, dim=-1)
             # sample from the distribution
             idx_next = torch.multinomial(probs, num_samples=1)
-            # append sampled index to the running sequence and continue
+            # append sampled index to the running sequence
             idx = torch.cat((idx, idx_next), dim=1)
```
Comment (danbraunai) [inline simple_stories_train/models/llama.py:428]: Should this be dim=-1 so it can handle the case with and without a batch index? I know you didn't right this, just spotted it and found it confusing.

File: simple_stories_train/models/llama.py [hunk 3]
```diff
@@ -400,6 +400,7 @@ def generate(
         max_new_tokens: int,
         temperature=1.0,
         top_k: int | None = None,
+        eos_token_id: int | None = None,  # Add EOS token parameter
```
Comment (danbraunai) [inline simple_stories_train/models/llama.py:403]: Comment not needed

File: tests/test_llama_implementation.py [hunk 4]
```diff
@@ -109,3 +109,45 @@ def test_local_pt_model_loading() -> None:
...
             torch.testing.assert_close(original_output, loaded_output)
             assert original_output.shape == (1, 1, config.vocab_size)
+
+
+def test_generate() -> None:
+    """Test the generate method including EOS token functionality for a single sequence"""
+    # Create a simple config for testing
+    config = LlamaConfig(
+        block_size=128,
+        vocab_size=300,
```
Comment (danbraunai) [inline tests/test_llama_implementation.py:119]: With a vocab size this large, you're very unlikely to sample an eos_token_id. I'd probably write two tests:
1. eos_token_id is None.
2. vocab_size is tiny, e.g. 2, and you generate 10 new tokens and ensure that it stops after eos_token_id. (Note that you might not want this if you're changing the functionality to not stop after eos_token_id if you're handling batch generation.

**Review by danbraunai — CHANGES_REQUESTED** [review]
Thanks! Added some comments.

File: simple_stories_train/models/llama.py (hunk 1 above)
Comment (chandanms) [reply simple_stories_train/models/llama.py:431]: My bad. I did not consider batch generation into account. I misread the function to be a single batch only. Let me see if I can rewrite it to handle batch generation. Thank you for pointing it out.

File: simple_stories_train/models/llama.py (hunk 2 above)
Comment (chandanms) [reply simple_stories_train/models/llama.py:452]: I think both point to the same thing ie sequence length but I feel `dim=1` is more explicit than `dim=-1`

Comment (danbraunai) [reply simple_stories_train/models/llama.py:452]: But if there is no batch index then won't this raise an error with dim=1? I think this generate needs a test for batch and non batch inputs.

**Review by danbraunai — CHANGES_REQUESTED** [review]
I think there's some confusion about "single input" vs "batched input" to the generate method. A single input would be `torch.tensor([3,9,0])`, and a batched input would be `torch.tensor([[3,9,0]])`.

The current code assumes that all inputs have a batch index, but I think the original code was intending to allow one to pass in a single input that doesn't have a batch index.

I think the best way to proceed is to just check the number of dimensions at the very start of the method and force it to be a batched input. The rest of the code should be fine.

File: tests/test_llama_implementation.py [hunk 5]
```diff
@@ -109,3 +109,108 @@ def test_local_pt_model_loading() -> None:
...
+        for seq_idx in range(output.shape[0]):
+            eos_positions = (output[seq_idx] == eos_token_id).nonzero()
+            if len(eos_positions) > 0:
+                first_eos_pos = eos_positions[0].item()
+                assert torch.all(output[seq_idx, first_eos_pos:] == eos_token_id), (
+                    "All tokens after first EOS should be EOS tokens"
+                )
+
+    # Test single sequence
+    single_input = torch.tensor([[0, 1, 2]], dtype=torch.long)
```
[the remark this answers is not in the document]
Comment (chandanms) [reply tests/test_llama_implementation.py]: Ah, sorry for the confusion. I had only worked with batch input before and misinterpreted the type hint of the method. My bad. I pushed the changes where it checks the number of dimensions at the start and changed the test cases too. Let me know if you want me to rewrite it based on original intention

**Review by danbraunai — APPROVED** [review]
Looks good! I'll merge this in. Thanks.
